app/models/mapping_rule.rb

```diff
@@ -11,94 +11,51 @@ class << self
     #
     # `dist/` serialization
 
-    def legacy_as_json(mapping_rules, version:)
+    def as_json(mapping_rules, version:)
       {
         "version" => version,
-        "mappings" => mapping_blocks_as_json(mapping_rules),
+        "mappings" => integration_blocks.map { |source, target|
+          mapping_blocks_as_json(source, target, mapping_rules)
+        },
       }
     end
 
     private
 
-    # TODO: Make this normative
-    def mapping_blocks_as_json(mapping_rules)
-      mapping_rule_blocks = Integration.all.pluck(:id, :available_versions).flat_map do |id, versions|
-        [id].product(versions, [true, false])
-      end.filter_map do |integration_id, version, from_shopify|
-        rules = mapping_rules.where(integration_id:, from_shopify:)
-        rules = if from_shopify
-          rules.where(output_version: version)
-        else
-          rules.where(input_version: version)
-        end
-        rules if rules.any?
-      end
-
-      mapping_rule_blocks.map do |mapping_rules|
-        rules = MappingBuilder.simple_mapping(mapping_rules:)
-        {
-          "input_taxonomy" => mapping_rules.first.input_version,
-          "output_taxonomy" => mapping_rules.first.output_version,
-          "rules" => rules&.filter_map { serialize_rule(_1) },
-        }
+    def integration_blocks
+      Integration.all.pluck(:id, :available_versions).flat_map do |id, versions|
+        source, destination = where(integration_id: id).pluck(:input_version, :output_version).uniq
       end
     end
 
-    def serialize_rule(mapping)
-      return if mapping.nil?
-
-      if mapping[:input][:attributes].present?
-        mapping[:input][:attributes] = mapping[:input][:attributes].map do |attribute|
-          {
-            name: Attribute.find(attribute[:name]).gid,
-            value: attribute[:value].nil? ? nil : Value.find(attribute[:value]).gid,
-          }
-        end
-      end
-
+    def mapping_blocks_as_json(input_version, output_version, mapping_rules)
       {
-        "input" => mapping[:input],
-        "output" => mapping[:output],
+        "input_taxonomy" => input_version,
+        "output_taxonomy" => output_version,
+        "rules" => mapping_rules.select {
+          _1.input_version == input_version && _1.output_version == output_version
+        }.map(&:rule_as_json),
       }
     end
   end
 
-  def apply(output_acc)
-    output_hash = output.payload.as_json
-    output_hash.delete_if { |_, value| value == [] }
-    output_acc.merge(output_hash) do |key, old_val, new_val|
-      intersection = old_val.intersection(new_val)
-      if !old_val.empty? && intersection.empty?
-        raise "Rules conflicted, please review the rules.
-          rule: #{as_json}, key: #{key}, old_val: #{old_val}, new_val: #{new_val}"
-      end
-
-      intersection
-    end
+  def rule_as_json
```

Comment:
This isn't a class method - I can definitely change it to `as_json`.

Reply:
I don't spend enough time doing class-based programming I guess because my brain will skip a couple frames figuring out which method is called. Maybe it's just Ruby's `<< self` wrapping the class methods. 🤷  

This is a TIL more than feedback for the record.